opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1509,54 +1505,51 @@ namespace Opm {
         const bool network_changed = comm.sum(local_network_changed);
         const double network_imbalance = comm.max(local_network_imbalance);
 
-        std::set<std::string> switched_wells;
-
-        if (checkGroupControls) {
-
-            // Check group individual constraints.
-            const int nupcol = schedule()[episodeIdx].nupcol();
-            // don't switch group control when iterationIdx > nupcol
-            // to avoid oscilations between group controls
-            if (iterationIdx <= nupcol) {
-                const Group& fieldGroup = schedule().getGroup("FIELD", episodeIdx);
-                updateGroupControls(fieldGroup, deferred_logger, episodeIdx, iterationIdx);
-            }
-            // Check wells' group constraints and communicate.
-            bool changed_well_group = false;
-            for (const auto& well : well_container_) {
-                const auto mode = WellInterface<TypeTag>::IndividualOrGroup::Group;
-                const bool changed_well = well->updateWellControl(ebosSimulator_, mode, this->wellState(), this->groupState(), deferred_logger);
-                if (changed_well) {
-                    switched_wells.insert(well->name());
-                    changed_well_group = changed_well || changed_well_group;
-                }
+        bool changed_well_group = false;
+        // Check group individual constraints.
+        const int nupcol = schedule()[episodeIdx].nupcol();
+        // don't switch group control when iterationIdx > nupcol
+        // to avoid oscilations between group controls
+        if (iterationIdx <= nupcol) {
+            const Group& fieldGroup = schedule().getGroup("FIELD", episodeIdx);
+            changed_well_group = updateGroupControls(fieldGroup, deferred_logger, episodeIdx, iterationIdx);
+        }
+        // Check wells' group constraints and communicate.
+        bool changed_well_to_group = false;
+        for (const auto& well : well_container_) {
+            const auto mode = WellInterface<TypeTag>::IndividualOrGroup::Group;
+            const bool changed_well = well->updateWellControl(ebosSimulator_, mode, this->wellState(), this->groupState(), deferred_logger);
+            if (changed_well) {
+                changed_well_to_group = changed_well || changed_well_to_group;
             }
-            changed_well_group = comm.sum(changed_well_group);
-            if (changed_well_group)
-                updateAndCommunicate(episodeIdx, iterationIdx, deferred_logger);
+        }
+
+        changed_well_to_group = comm.sum(changed_well_to_group);
+        if (changed_well_to_group) {
```

Comment:
To be clear: you only seek to update and communicate if updateWellControl() returned true for a well, irrespective of whether updateGroupControls() returned true or false?

Reply:
Yes it should return true if either a well or a group control is changed. So If a well control is changed we don't case whether a group also changed.